[Arnaxx54] iP

arnav-goel05 opened the pull request:
# DrBrownBot

Welcome to **DrBrown**, a chatbot inspired by _Back to the Future_! :tophat: DrBrown is a smart assistant that helps you manage your tasks efficiently. 

> "Great Scott! This chatbot is a fantastic way to organize your time!" - Dr. Emmett Brown

## How to Use DrBrown

1. **Create a task**: 
   - **Todo**: `todo {description}`
   - **Deadline**: `deadline {description} /by {date time}`
   - **Event**: `event {description} /from {date time} /to {date time}`
2. **View tasks**: Use the `list` command to display all your tasks.
3. **Manage tasks**:
   - **Mark/Unmark**: Use `mark {index}` to mark a task as done or `unmark {index}` to mark it as undone.
   - **Delete**: Use `delete {index}` to remove a task from the list.
4. **Search tasks**: Use the `find {string}` command to search tasks by description matching the provided string.
5. **End the program**: Simply type `bye` to exit the chatbot.

## Features of DrBrown

- [x] Allows you to manage tasks (todo, deadline, event)
- [x] List out all tasks
- [x] Delete tasks
- [x] Mark tasks as done or undone
- [x] Search task descriptions by keyword
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

For more information, check out the project on [GitHub](https://github.com/Arnaxx54/ip).

```java
public static void main(String[] args) {
    new DrBrown("data/DrBrown.txt").run();
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

arnav-goel05 committed:
Update name and add greeting/exit messages

arnav-goel05 committed:
Update ability to echo user response if not bye

arnav-goel05 committed:
Add support for list

arnav-goel05 committed:
Add support for tasks

arnav-goel05 committed:
Add support for marking and unmarking tasks

arnav-goel05 committed:
Add support for todo, event and deadlines

arnav-goel05 committed:
Update testing files

arnav-goel05 committed:
Handle errors

arnav-goel05 committed:
Add delete

arnav-goel05 committed:
Update save functionality

arnav-goel05 committed:
Fix functionality for adding tasks to file

arnav-goel05 committed:
Implement ability to handle date time formatting

arnav-goel05 committed:
Merge branch 'branch-Level-8'

arnav-goel05 committed:
Transfer UI functionality to UI.java

arnav-goel05 committed:
Transfer parsing code to Parser.java

arnav-goel05 committed:
Fix error handling and data-time parsing

arnav-goel05 committed:
Extract code to individual classes to follow OOP

arnav-goel05 committed:
Merge branch 'branch-A-MoreOOP' implementing A-MoreOOP

arnav-goel05 committed:
Divide classes into packages

arnav-goel05 committed:
Update test files

arnav-goel05 committed:
Add gradle support

arnav-goel05 committed:
Delete old file

arnav-goel05 committed:
Use gradle

arnav-goel05 committed:
Add JUnit test

arnav-goel05 committed:
Add JavaDoc comments

arnav-goel05 committed:
Merge branch implementing JavaDoc comments

arnav-goel05 committed:
Comply with coding standards

arnav-goel05 committed:
Resolved merge conflicts and merged branch-A-CodingStandard

arnav-goel05 committed:
Implement find

arnav-goel05 committed:
Resolve merge conflicts and merge branch-Level-9

arnav-goel05 committed:
Fix OOP encapsulation

gohqingkhang commented on src/main/java/drbrown/DrBrown.java:
```diff
@@ -0,0 +1,70 @@
+     * Constructs a new DrBrown instance with the specified file path for storage.
+     * Initializes the UI, loads the task list from storage, and handles any exceptions during loading.
+     *
+     * @param filePath the file path to load and save tasks
```
Should there be a punctation after each parameter description?

gohqingkhang commented on src/main/java/drbrown/DrBrown.java:
```diff
@@ -0,0 +1,70 @@
+     * The main method to start the DrBrown application.
+     * It creates a new instance of DrBrown with the specified file path and runs the application.
+     *
+     * @param args command-line arguments (not used)
```
Similar issue as above with the punctuation after each parameter description :)

gohqingkhang commented on src/main/java/drbrown/command/AddCommand.java:
```diff
@@ -0,0 +1,48 @@
+package drbrown.command;
+
+import drbrown.utils.Storage;
+import drbrown.task.Task;
```
Should import for Task class be placed here? Maybe can consider lexicographical order?

gohqingkhang commented on src/main/java/drbrown/command/AddCommand.java:
```diff
@@ -0,0 +1,48 @@
+     */
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        ui.showTaskCreation(task);
```
Well done in passing job of printing message to the ui!

gohqingkhang commented on src/main/java/drbrown/utils/Parser.java:
```diff
@@ -0,0 +1,130 @@
+package drbrown.utils;
+
+import drbrown.command.*;
```
Maybe you should list out all the imports explicitly instead of using *?

gohqingkhang commented on src/main/java/drbrown/utils/Parser.java:
```diff
@@ -0,0 +1,130 @@
+                    return new AddCommand(todo);
+                } catch (ArrayIndexOutOfBoundsException e) {
+                    throw new DrBrownException("Great Scott! You can't add a to-do without a description!\n\nUse the format: todo {description}");
+                }
```
Maybe the indentation for case clause should be one tab less? Same applies for the other case clauses below :)

gohqingkhang commented on src/main/java/drbrown/utils/Parser.java:
```diff
@@ -0,0 +1,130 @@
+ */
+public class Parser {
+
+    private static final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM dd yyyy HH:mm");
```
Good use of constant for formatter! However, maybe can you capitalized naming instead?
```suggestion
    private static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("MMM dd yyyy HH:mm");
```

gohqingkhang reviewed (commented):
Overall, very well done for JavaDoc Comments! Just a few nits with coding standards :)

arnav-goel05 committed:
Update code to follow style conventions

arnav-goel05 committed:
Update code to match checkstyle convention

arnav-goel05 committed:
Update code to match checkstyle convention

arnav-goel05 committed:
Merge checkstyle branch into master

yugan01 commented on src/main/java/drbrown/command/Command.java:
```diff
@@ -0,0 +1,30 @@
+package drbrown.command;
+
+import drbrown.utils.DrBrownException;
+import drbrown.utils.Storage;
+import drbrown.utils.TaskList;
+import drbrown.utils.Ui;
+
+/**
+ * Represents an abstract command that can be executed.
+ * All specific command types should extend this class and implement its abstract methods.
+ */
+public abstract class Command {
+
+    /**
+     * Executes the command with the specified task list, UI, and storage.
+     *
+     * @param tasks   The TaskList containing the current tasks.
+     * @param ui      The UI object to interact with the user.
+     * @param storage The Storage object to save and load tasks.
+     * @throws DrBrownException If there is an error executing the command.
+     */
+    public abstract void executeCommand(TaskList tasks, Ui ui, Storage storage) throws DrBrownException;
+
+    /**
+     * Determines whether the command exits the program.
+     *
+     * @return true if the command exits the program, false otherwise.
+     */
+    public abstract boolean shouldExit();
+}
```
Is there any reason why this class is abstract and not made as a interface as all methods are abstract?

yugan01 commented on src/main/java/drbrown/command/FindCommand.java:
```diff
@@ -0,0 +1,53 @@
+     */
+    @Override
+    public void executeCommand(TaskList tasks, Ui ui, Storage storage) throws DrBrownException {
+        if (tasks.getCount() == 0) {
+            throw new DrBrownException("Wait a minute, Doc! There's nothing here! We can't go "
+                    + "anywhere until you add something to the list!");
+        }
```
I really like how creative and personal these error messages are!

yugan01 commented on src/main/java/drbrown/task/Event.java:
```diff
@@ -0,0 +1,64 @@
+     */
+    @Override
+    public String toString() {
+        return "[E][" + (getStatus() ? "X" : " ") + "] " + getDescription()
```
Could this be abstracted out in Task.java 

In Task.java
```
public String toString() {
          return getStatus() ? "X" : " ") + "] " + getDescription();
}
```
and in ur subclasses, it could just be 
```
return "[E][" + super.toString() + .... ;

yugan01 reviewed (commented):
LGTM, I really enjoyed reading through your code as the Ui messages were very creative. Perhaps changing Command to an interface might make the code look cleaner and using the Task toString method to abstract the getStatus() and getDescription() might be beneficial as well.

arnav-goel05 committed:
Add graphical user interface

arnav-goel05 committed:
Merge branch-Level-10 into master

arnav-goel05 committed:
Add assertions to validate

To ensure arguments are valid and expected criteria is met. They help to catch errors early on

Validate function arguments, checking for null and whether value is within acceptable range

To catch any bugs created by the programmer and improve robustness of the programme

arnav-goel05 committed:
Improve code quality

Refactored Parser class to follow SLAP

Updated java docs

Simplified main code logic for easy understanding

The commit makes code more readable and understandable

arnav-goel05 committed:
Merge pull request #1 from Arnaxx54/branch-A-Assertions

Add assertions to validate

arnav-goel05 committed:
Resolve merge conflict in branch-A-CodeQuality

arnav-goel05 committed:
Update assert statements in parsing

arnav-goel05 committed:
Add feature to tag priority to tasks

arnav-goel05 committed:
Merge branch-C-Priority to master branch

arnav-goel05 committed:
Add Git CI workflows for automated testing and deployment

Integrated Continuous Integration (CI) workflows using GitHub Actions to automate the testing and deployment processes

This addition enhances the development cycle by providing continuous feedback, maintaining code quality, and facilitating faster deployments

arnav-goel05 committed:
Fix style errors

arnav-goel05 committed:
Updata javadocs

arnav-goel05 committed:
Fix style errors

arnav-goel05 committed:
Edit gradle.yml file

arnav-goel05 committed:
Create common parsing abstract class

arnav-goel05 committed:
Follow SLAP

arnav-goel05 committed:
Improve abstraction

arnav-goel05 committed:
Improve abstraction

arnav-goel05 committed:
Improve abstraction

arnav-goel05 committed:
Improve GUI and create user guide

arnav-goel05 committed:
Update JavaDocs

arnav-goel05 committed:
Revise user guide

arnav-goel05 committed:
Update readme docs with detailed usage instructions

arnav-goel05 committed:
Revise README for 'Back to the Future' Java Chatbot
